curves/src/pasta/wasm_friendly/backend9.rs

```diff
@@ -0,0 +1,213 @@
+    p[2] |= (pa[6] as u64) << 46;
+    p[3] = (pa[6] as u64) >> 18;
+    p[3] |= (pa[7] as u64) << 11;
+    p[3] |= (pa[8] as u64) << 40;
+    p
+}
+
+pub trait FpConstants: Send + Sync + 'static + Sized {
+    const MODULUS: B;
+    const MODULUS64: B64 = {
+        let mut modulus64 = [0u64; 9];
+        let modulus = Self::MODULUS;
+        let mut i = 0;
+        while i < 9 {
+            modulus64[i] = modulus[i] as u64;
+            i += 1;
+        }
+        modulus64
+    };
+
+    /// montgomery params
+    /// TODO: compute these
+    const R: B; // R = 2^261 mod modulus
+    const R2: B; // R^2 mod modulus
+    const MINV: u64; // -modulus^(-1) mod 2^29, as a u64
+}
+
+#[inline]
+fn gte_modulus<FpC: FpConstants>(x: &B) -> bool {
+    for i in (0..9).rev() {
+        // don't fix warning -- that makes it 15% slower!
+        #[allow(clippy::comparison_chain)]
+        if x[i] > FpC::MODULUS[i] {
+            return true;
+        } else if x[i] < FpC::MODULUS[i] {
+            return false;
+        }
+    }
+    true
+}
+
+// TODO performance ideas to test:
+// - unroll loops
+// - introduce locals for a[i] instead of accessing memory multiple times
+// - only do 1 carry pass at the end, by proving properties of greater-than on uncarried result
+// - use cheaper, approximate greater-than check a[8] > Fp::MODULUS[8]
+pub fn add_assign<FpC: FpConstants>(x: &mut B, y: &B) {
+    let mut tmp: u32;
+    let mut carry: i32 = 0;
+
+    for i in 0..9 {
+        tmp = x[i] + y[i] + (carry as u32);
+        carry = (tmp as i32) >> SHIFT;
+        x[i] = tmp & MASK;
+    }
+
+    if gte_modulus::<FpC>(x) {
+        carry = 0;
+        #[allow(clippy::needless_range_loop)]
+        for i in 0..9 {
+            tmp = x[i].wrapping_sub(FpC::MODULUS[i]) + (carry as u32);
+            carry = (tmp as i32) >> SHIFT;
+            x[i] = tmp & MASK;
+        }
+    }
+}
+
+#[inline]
+fn conditional_reduce<FpC: FpConstants>(x: &mut B) {
+    if gte_modulus::<FpC>(x) {
+        #[allow(clippy::needless_range_loop)]
+        for i in 0..9 {
+            x[i] = x[i].wrapping_sub(FpC::MODULUS[i]);
+        }
+        #[allow(clippy::needless_range_loop)]
+        for i in 1..9 {
+            x[i] += ((x[i - 1] as i32) >> SHIFT) as u32;
+        }
+        #[allow(clippy::needless_range_loop)]
+        for i in 0..8 {
+            x[i] &= MASK;
+        }
+    }
+}
+
+/// Montgomery multiplication
+pub fn mul_assign<FpC: FpConstants>(x: &mut B, y: &B) {
+    // load y[i] into local u64s
+    // TODO make sure these are locals
+    let mut y_local = [0u64; 9];
+    for i in 0..9 {
+        y_local[i] = y[i] as u64;
+    }
+
+    // locals for result
+    let mut z = [0u64; 8];
+    let mut tmp: u64;
+
+    // main loop, without intermediate carries except for z0
+    #[allow(clippy::needless_range_loop)]
+    for i in 0..9 {
+        let xi = x[i] as u64;
+
+        // compute qi and carry z0 result to z1 before discarding z0
+        tmp = xi * y_local[0];
```

Comment:
```suggestion
        tmp = (xi * y_local[0]) + z[0];
```

This is what's missing from the javascript implementation:
https://github.com/mitschabaude/montgomery/blob/4f1405073aef60b469a0fcd45ff9166f8de4a4bf/src/wasm/multiply-montgomery.ts#L83-L85

Reply:
great catch!